File: algorithms/js/200.岛屿数量.js
/*
 * @lc app=leetcode.cn id=200 lang=javascript
 *
 * [200] 岛屿数量
 *
 * https://leetcode-cn.com/problems/number-of-islands/description/
 *
 * algorithms
 * Medium (45.29%)
 * Likes:    206
 * Dislikes: 0
 * Total Accepted:    22.6K
 * Total Submissions: 49.9K
 * Testcase Example:  '[["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]'
 *
 * 给定一个由 '1'（陆地）和
 * '0'（水）组成的的二维网格，计算岛屿的数量。一个岛被水包围，并且它是通过水平方向或垂直方向上相邻的陆地连接而成的。你可以假设网格的四个边均被水包围。
 *
 * 示例 1:
 *
 * 输入:
 * 11110
 * 11010
 * 11000
 * 00000
 *
 * 输出: 1
 *
 *
 * 示例 2:
 *
 * 输入:
 * 11000
 * 11000
 * 00100
 * 00011
 *
 * 输出: 3
 *
 *
 */

// @lc code=start
function dfs(x, y, grid) {
  if (
    x < 0 ||
    y < 0 ||
    x >= grid.length ||
    y >= grid[x].length ||
    grid[x][y] === "0"
  )
    return;
  grid[x][y] = "0";
  dfs(x, y - 1, grid);
  dfs(x, y + 1, grid);
  dfs(x - 1, y, grid);
  dfs(x + 1, y, grid);
}
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  let num = 0;
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (grid[i][j] === "1") {
        dfs(i, j, grid);
        num++;
      }
    }
  }
  return num;
};
// @lc code=end

File: algorithms/golang/200.岛屿数量.go
/*
 * @lc app=leetcode.cn id=200 lang=golang
 *
 * [200] 岛屿数量
 */

// @lc code=start

var Directions [][]int = [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

/**
 * DFS
 */
func numIslands_1(grid [][]byte) int {
	ret := 0
	row, col := len(grid), len(grid[0])
	var dfs func(row, col int)
	dfs = func(x, y int) {
		if x >= row || y >= col || x < 0 || y < 0 || grid[x][y] == '0' {
			return
		}
		grid[x][y] = '0'
		for _, direction := range Directions {
			x1, y1 := x+direction[0], y+direction[1]
			dfs(x1, y1)
		}
	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if grid[i][j] == '1' {
				ret++
				dfs(i, j)
			}
		}
	}
	return ret
}

/**
 * BFS
 */
func numIslands(grid [][]byte) int {
	ret := 0
	row, col := len(grid), len(grid[0])
	var bfs func(row, col int)
	bfs = func(x, y int) {
		q := [][]int{{x, y}}
		for len(q) > 0 {
			size := len(q)
			for i := 0; i < size; i++ {
				posX, posY := q[i][0], q[i][1]
				if posX >= row || posY >= col || posX < 0 || posY < 0 || grid[posX][posY] == '0' {
					continue
				}
				grid[posX][posY] = '0'
				for _, direction := range Directions {
					x1, y1 := posX+direction[0], posY+direction[1]
					q = append(q, []int{x1, y1})
				}
			}
			q = q[size:]
		}
	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if grid[i][j] == '1' {
				ret++
				bfs(i, j)
			}
		}
	}
	return ret
}

// @lc code=end
